kedro/extras/datasets/video/video_dataset.py

```diff
@@ -0,0 +1,301 @@
+            raise IndexError()
+
+        if index != self._index:
+            self._cap.set(cv2.CAP_PROP_POS_FRAMES, index)
+        self._index = index + 1  # Next frame to decode after this
+        ret, frame_bgr = self._cap.read()
+        if not ret:
+            raise IndexError()
+
+        height, width = frame_bgr.shape[:2]
+        return PIL.Image.frombuffer(  # Convert to PIL image with RGB instead of BGR
+            "RGB", (width, height), frame_bgr, "raw", "BGR", 0, 0
+        )
+
+    def _get_length(self) -> int:
+        # OpenCV's frame count might be an approximation depending on what
+        # headers are available in the video file
+        length = int(round(self._cap.get(cv2.CAP_PROP_FRAME_COUNT)))
+        if length >= 0:
+            return length
+
+        # On some video files it fails, we can count the frames but it will be
+        # very slow... It also seems like that opencv has issues to seek in those
+        # videos...
+        raise ValueError("Failed to load video since number of frames can't be infered")
+
+
+class SequenceVideo(AbstractVideo):
+    """A video object read from an indexable sequence of frames"""
+
+    def __init__(
+        self, frames: Sequence[PIL.Image.Image], fps: float, fourcc: str = "mp4v"
+    ) -> None:
+        self._n_frames = len(frames)
+        self._frames = frames
+        self._fourcc = fourcc
+        self._size = frames[0].size
+        self._fps = fps
+
+    @property
+    def fourcc(self) -> str:
+        return self._fourcc
+
+    @property
+    def fps(self) -> float:
+        return self._fps
+
+    @property
+    def size(self) -> Tuple[int, int]:
+        return self._size
+
+    def __getitem__(self, index: Union[int, slice]):
+        if isinstance(index, slice):
+            return SlicedVideo(self, index)
+        return self._frames[index]
+
+
+class GeneratorVideo(AbstractVideo):
+    """A video object with frames yielded by a generator"""
+
+    def __init__(
+        self,
+        frames: Generator[PIL.Image.Image, None, None],
+        length,
+        fps: float,
+        fourcc: str = "mp4v",
+    ) -> None:
+        self._n_frames = length
+        self._gen = more_itertools.peekable(frames)
+        self._fourcc = fourcc
+        self._size = self._gen.peek().size
+        self._fps = fps
+
+    @property
+    def fourcc(self) -> str:
+        return self._fourcc
+
+    @property
+    def fps(self) -> float:
+        return self._fps
+
+    @property
+    def size(self) -> Tuple[int, int]:
+        return self._size
+
+    def __getitem__(self, index: Union[int, slice]):
+        raise NotImplementedError("Underlying video is a generator")
+
+    def __next__(self):
+        return next(self._gen)
+
+    def __iter__(self):
+        return self
+
+
+class VideoDataSet(AbstractDataSet):
+    """``VideoDataSet`` loads / save video data from a given filepath as sequence
+    of PIL.Image.Image using OpenCV.
+
+    Example:
```

Comment:
Can you add an example of how to define this in yaml in the catalog as well?

Reply:
Done.